certbot/src/certbot/_internal/san.py

```diff
@@ -0,0 +1,164 @@
+        return hash(self.dns_name)
+
+    def __repr__(self) -> str:
+        return 'DNS(%s)' % self.dns_name
+
+    def __eq__(self, other: Any) -> bool:
+        match other:
+            case DNSName():
+                return self.dns_name == other.dns_name
+            case IPAddress():
+                return False
+            case _:
+                raise TypeError(f"DNSName SAN compared to non-SAN: {type(other)}")
+
+    def is_wildcard(self) -> bool:
+        """Return True if this DNS name is a wildcard."""
+        return self.dns_name.startswith('*.')
+
+class IPAddress(SAN):
+    """An IP address (IPv4 or IPv6).
+
+    Validated upon construction.
+    """
+    def __init__(self, ip_address: str) -> None:
+        self.ip_address = ipaddress.ip_address(ip_address)
+
+    def __str__(self) -> str:
+        return str(self.ip_address)
+
+    def __hash__(self) -> int:
+        return hash(self.ip_address)
+
+    def __repr__(self) -> str:
+        return 'IP(%s)' % self.ip_address
+
+    def __eq__(self, other: Any) -> bool:
+        match other:
+            case IPAddress():
+                return self.ip_address == other.ip_address
+            case DNSName():
+                return False
+            case _:
+                raise TypeError(f"IPAddress SAN compared to non-SAN: {type(other)}")
+
+    def is_wildcard(self) -> bool:
+        """Always False."""
+        return False
+
+def split(sans: Iterable[SAN]) -> tuple[list[DNSName], list[IPAddress]]:
+    """Split a list of SANs into a list of DNSNames and one of IPAddress, in that order."""
+    domains = []
+    ip_addresses = []
+    for s in sans:
+        match s:
+            case IPAddress():
+                ip_addresses.append(s)
+            case DNSName():
+                domains.append(s)
+            case _:
+                raise TypeError(f"SAN of type {type(s)}")
+    return domains, ip_addresses
+
+def display(sans: Iterable[SAN]) -> str:
+    """Return the list of SANs in string form, separated by comma and space."""
+    return ", ".join(map(str, sans))
+
+def from_x509(subject: x509.Name, exts: x509.Extensions) -> list[SAN]:
+    """Get all DNS names and IP addresses, plus the first Common Name from subject.
+
+    The CN will be first in the list, if present. It will always be interpreted
+    as a DNS name.
+
+    :param subject: Name of the x509 object, which may include Common Name
+    :type subject: `cryptography.x509.Name`
+    :param exts: Extensions of the x509 object, which may include SANs
+    :type exts: `cryptography.x509.Extensions`
+
+    :returns: List of DNS Subject Alternative Names and first Common Name
+    :rtype: `list` of `SAN`
+    """
+    dns_names, ip_addresses = acme_crypto_util.get_identifiers_from_x509(subject, exts)
+    sans: list[SAN] = []
+    for dns_name in dns_names:
+        sans.append(DNSName(dns_name))
+    for ip_address in ip_addresses:
+        sans.append(IPAddress(ip_address))
+    return sans
```

Comment:
looks like in the one place this function is used, we immediately then split them back into two lists. seems like it'd be better in that case to just return two lists, unless there's a specific reason I'm not seeing to return them as one list?

Reply:
Good point. Turned it into two lists. Technically there's one place in #10468 that doesn't immediately split `RenewableCert.sans()`. But it's fine to have that function combine the two lists. And then every caller of `.sans()` gets the combined list in a convenient way.

Reply:
Ah, I'd actually missed that it was called in `RenewableCert.sans()` somehow, not sure what happened there. But yeah, seems reasonable to combine it there imo.